### PR TITLE
Remove queued state, and centralize queue logic.

### LINKED ISSUE #4013: Implement task state rationalization proposal
Complete task state rationalization as described in [this proposal](https://cylc.github.io/cylc-admin/proposal-state-names.html)

- [x] (waiting, queued, runahead) -> waiting (finished by #4035)
- [x] ready -> preparing  #4014
- [x] retrying, submit-retrying: re-implement as xtrigger  #3423
- [x] held: re-implement as attribute  #3230

Note we decided not to do the proposed trivial renaming (succeeded -> run-succeeded, etc.).

### PR DESCRIPTION
**Pull request by hjoliver**
Supersedes #4035 
Close #4013

- remove obsolete *queued* task state
- centralize queuing logic, and re-implement only our "classic" queues, using actual queues and simple push/pop
- abstract out the queuing API, to make it easier to try other queuing systems in future

(I backed out of the new "overlapping queues" in order to re-implement classic queues simply and efficiently; I'll save the overlapping implementation to experiment with later).

@oliver-sanders and @kinow - apologies for the double review. But it should be easy: this one is basically the same but simpler: no overlapping option, no confusing single-queue-with-queue-jumping implementation, plus a new abstract base class.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required. (Covered by general task state documentation overhaul)
- [x] No dependency changes.

### CODE REVIEW

File: cylc/flow/network/schema.py [hunk 1]
```diff
@@ -309,7 +311,7 @@ class SortArgs(InputObjectType):
     'exids': List(ID, default_value=[]),
     'states': List(String, default_value=[]),
     'exstates': List(String, default_value=[]),
-    'is_held': Boolean(),
+    'is_queued': Boolean(),
```
Comment (kinow) [inline cylc/flow/network/schema.py:314]: @hjoliver the GitHub UI shows `NODES_EDGES_ARGS` with `is_held` **and new `is_queued`. But `NODES_EDGES_ARGS_ALL` with `is_queued` replacing `is_held`. Is that correct?

File: cylc/flow/scheduler.py [hunk 2]
```diff
@@ -1187,28 +1191,45 @@ def run_event_handlers(self, event, reason):
...
+        self.pre_submit_tasks = [
+            itask for itask in self.pre_submit_tasks if
+            itask.waiting_on_job_prep
+        ]
+
         if self.stop_mode is None and self.auto_restart_time is None:
-            itasks = self.pool.get_ready_tasks()
-            if itasks:
+            # Add newly released tasks to those still preparing.
+            from datetime import datetime
```
Comment (kinow) [inline cylc/flow/scheduler.py:1216]: I wonder if this is preventing `datetime` from being imported or not.

```bash
(venv) kinow@ranma:~/Development/python/workspace/cylc-flow$ python
Python 3.8.3 (default, May 19 2020, 18:47:26) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from cylc.flow.scheduler import Scheduler
>>> import sys
>>> [mod for mod in sys.modules.keys() if 'datetime' in mod]
['_datetime', 'datetime', 'metomi.isodatetime', 'metomi.isodatetime.timezone', 'metomi.isodatetime.exceptions', 'metomi.isodatetime.parser_spec', 'metomi.isodatetime.dumpers', 'metomi.isodatetime.data', 'metomi.isodatetime.parsers', 'graphene.types.datetime']
>>> 
```

I think importing `Scheduler` is loading `datetime` too transitively? (I couldn't find which module is causing that to load). But that's not loaded by default.

```bash
(venv) kinow@ranma:~/Development/python/workspace/cylc-flow$ python
Python 3.8.3 (default, May 19 2020, 18:47:26) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> [mod for mod in sys.modules.keys() if 'datetime' in mod]
[]
>>> 

```

File: cylc/flow/task_pool.py [hunk 3]
```diff
@@ -585,15 +577,14 @@ def remove(self, itask, reason=""):
             try:
                 del self.pool[itask.point][itask.identity]
             except KeyError:
-                # Not in main pool (forced spawn uses temporary non-pool tasks)
+                # TODO: not possible under SoD?
```
Comment (kinow) [inline cylc/flow/task_pool.py:580]: :+1:

File: cylc/flow/scheduler.py (hunk 2 above)
Comment (hjoliver) [reply cylc/flow/scheduler.py:1216]: oops, good spotting - that's a debug left-over

File: cylc/flow/network/schema.py (hunk 1 above)
Comment (hjoliver) [reply cylc/flow/network/schema.py:314]: good spotting again :1st_place_medal: that's an error, supposed to have both in the schema.

**Review by kinow — APPROVED** [review]
Had finished the code review last Friday, but didn't have time to test it. Tested today with the sibling UI PR https://github.com/cylc/cylc-ui/pull/575, worked with no issues :+1: 

Screenshot from 2021-02-22 12-29-33

**Review by oliver-sanders — APPROVED** [review]

**Comment by hjoliver** [comment]
(Rebased and deconflicted; I'll merge if tests pass).

**Pull request merged**